README.md
# Youthweb Provider for OAuth 2.0 Client

[![Latest Version](https://img.shields.io/github/release/youthweb/oauth2-youthweb.svg)](https://github.com/youthweb/oauth2-youthweb/releases)
[![Software License GLPv3](http://img.shields.io/badge/License-GPLv3-brightgreen.svg)](LICENSE)
[![Build Status](http://img.shields.io/travis/youthweb/oauth2-youthweb.svg)](https://travis-ci.org/youthweb/oauth2-youthweb)
[![Coverage Status](https://coveralls.io/repos/youthweb/oauth2-youthweb/badge.svg?branch=master&service=github)](https://coveralls.io/github/youthweb/oauth2-youthweb?branch=master)
[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/youthweb/youthweb-api?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

This package provides [Youthweb OAuth 2.0 support](https://developer.youthweb.net/api_general_oauth2.html) for the PHP League's [OAuth 2.0 Client](https://github.com/thephpleague/oauth2-client).

## Installation

To install, use composer:

```
composer require youthweb/oauth2-youthweb
```

## Usage

Usage is the same as The League's OAuth client, using `\Youthweb\OAuth2\Client\Provider\Youthweb` as the provider.

### Authorization Code Flow

```php
$provider = new Youthweb\OAuth2\Client\Provider\Youthweb([
    'clientId'          => '{youthweb-client-id}',
    'clientSecret'      => '{youthweb-client-secret}',
    'redirectUri'       => 'https://example.org/callback-url',
    'apiVersion'        => '0.15', // optional
]);

if ( ! isset($_GET['code']) )
{
    // If we don't have an authorization code then get one
    $authUrl = $provider->getAuthorizationUrl();
    $_SESSION['oauth2state'] = $provider->getState();
    header('Location: '.$authUrl);
    exit;
}
// Check given state against previously stored one to mitigate CSRF attack
elseif ( empty($_GET['state']) || ($_GET['state'] !== $_SESSION['oauth2state']) )
{
    unset($_SESSION['oauth2state']);
    exit('Invalid state');
}
else
{
    // Try to get an access token (using the authorization code grant)
    $token = $provider->getAccessToken('authorization_code', [
        'code' => $_GET['code']
    ]);

    // Optional: Now you have a token you can look up a users profile data
    try
    {
        // We got an access token, let's now get the user's details
        $user = $provider->getResourceOwner($token);

        // Use these details to create a new profile
        printf('Hello %s!', $user->getNickname());
    }
    catch (Exception $e)
    {
        // Failed to get user details
        exit('Oh dear...');
    }

    // Use this to interact with an API on the users behalf
    echo $token->getToken();
}
```

### Managing Scopes

When creating your Youthweb authorization URL, you can specify the state and scopes your application may authorize.

```php
$options = [
    'state' => 'OPTIONAL_CUSTOM_CONFIGURED_STATE',
    'scope' => ['user:read', 'user:email'] // array or string
];

$authorizationUrl = $provider->getAuthorizationUrl($options);
```
If neither are defined, the provider will utilize internal defaults.

At the time of authoring this documentation, the following scopes are available with Youthweb-API 0.15:

- post:read
- post:write
- user:read
- user:email

See [here](https://developer.youthweb.net/api_general_scopes.html) for more information.

## Testing

``` bash
$ ./vendor/bin/phpunit
```

## Contributing

Please feel free to fork and sending Pull Requests. This project follows [Semantic Versioning 2](http://semver.org) and [PSR-2](https://www.php-fig.org/psr/psr-2/).

## Credits

- [All Contributors](https://github.com/youthweb/oauth2-youthweb/contributors)

## License

GPL3. Please see [License File](LICENSE) for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased](https://github.com/youthweb/oauth2-youthweb/compare/1.2.0...master)

## [1.2.0](https://github.com/youthweb/oauth2-youthweb/compare/1.1.0...1.2.0) - 2021-03-05

### Added

- Add support for PHP 7.4 and PHP 8.0

### Changed

- The default api version was bumped to 0.18
- Drop support for PHP 7.2 and 7.3

## [1.1.0](https://github.com/youthweb/oauth2-youthweb/compare/1.0.2...1.1.0) - 2019-10-01

### Added

- The api version is now configurable via `Youthweb\OAuth2\Client\Provider\Youthweb::__construct()`
- Added support for PHP 7.2 and 7.3

### Changed

- Update default api version to 0.15
- Drop support for PHP 5.6, 7.0 and 7.1
- Change Code Style to PSR-2

## [1.0.2](https://github.com/youthweb/oauth2-youthweb/compare/1.0.1...1.0.2) - 2017-06-04

### Changed

- Library uses Youthweb-API 0.9

## [1.0.1](https://github.com/youthweb/oauth2-youthweb/compare/1.0.0...1.0.1) - 2016-10-27

### Changed

- Fixed error handling for JSON API errors

## [1.0.0](https://github.com/youthweb/oauth2-youthweb/compare/2ab1e2eefcf5ef86de03995ce5e804b24a2b118d...1.0.0) - 2016-10-14

### Added

- Initial release


tests/Unit/Provider/YouthwebTest.php
<?php
/*
 * A PHP Library to simplify the OAuth2 process with youthweb.net.
 * Copyright (C) 2016-2019  Youthweb e.V. <info@youthweb.net>

 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Youthweb\OAuth2\Client\Tests\Unit\Provider;

class YouthwebTest extends \PHPUnit\Framework\TestCase
{
    protected $provider;

    protected function setUp(): void
    {
        $this->provider = new \Youthweb\OAuth2\Client\Provider\Youthweb([
            'clientId' => 'mock_client_id',
            'clientSecret' => 'mock_secret',
            'redirectUri' => 'none',
            'apiVersion' => '0.1',
        ]);
    }

    public function testAuthorizationUrl()
    {
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);
        parse_str($uri['query'], $query);

        $this->assertArrayHasKey('client_id', $query);
        $this->assertArrayHasKey('redirect_uri', $query);
        $this->assertArrayHasKey('state', $query);
        $this->assertArrayHasKey('scope', $query);
        $this->assertArrayHasKey('response_type', $query);
        $this->assertArrayHasKey('approval_prompt', $query);
        $this->assertNotNull($this->provider->getState());
    }

    public function testScopes()
    {
        $options = ['scope' => [uniqid(),uniqid()]];

        $url = $this->provider->getAuthorizationUrl($options);

        $this->assertStringContainsString(urlencode(implode(',', $options['scope'])), $url);
    }

    public function testGetAuthorizationUrl()
    {
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);

        $this->assertEquals('/auth/authorize', $uri['path']);
    }

    public function testGetBaseAccessTokenUrl()
    {
        $params = [];

        $url = $this->provider->getBaseAccessTokenUrl($params);
        $uri = parse_url($url);

        $this->assertEquals('/auth/access_token', $uri['path']);
    }

    public function testGetAccessToken()
    {
        $response = $this->createMock('Psr\Http\Message\ResponseInterface');
        $response->method('getBody')->willReturn('{"access_token":"mock_access_token", "scope":"user:read", "token_type":"bearer"}');
        $response->method('getHeader')->willReturn(['content-type' => 'json']);
        $response->method('getStatusCode')->willReturn(200);

        $client = $this->createMock('GuzzleHttp\ClientInterface');
        $client->expects($this->once())->method('send')->willReturn($response);
        $this->provider->setHttpClient($client);

        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);

        $this->assertEquals('mock_access_token', $token->getToken());
        $this->assertNull($token->getExpires());
        $this->assertNull($token->getRefreshToken());
        $this->assertNull($token->getResourceOwnerId());
    }

    public function testUserData()
    {
        $userId = rand(1000, 99999);
        $firstname = uniqid();
        $lastname = uniqid();
        $nickname = uniqid();
        $email = uniqid();

        $postResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
        $postResponse->method('getBody')->willReturn('access_token=mock_access_token&expires=3600&refresh_token=mock_refresh_token&otherKey={1234}');
        $postResponse->method('getHeader')->willReturn(['content-type' => 'application/x-www-form-urlencoded']);
        $postResponse->method('getStatusCode')->willReturn(200);

        $userResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
        $userResponse->method('getBody')->willReturn('{"data": {"type": "users","id": "' . $userId . '","attributes": {"username": "' . $nickname . '","first_name": "' . $firstname . '","last_name": "' . $lastname . '","email": "' . $email . '","birthday": "1988-03-05","created_at": "2006-01-01T21:00:00+01:00","last_login": "2016-01-01T22:00:00+02:00","zip": "12345","city": "Jamestown","description_jesus": "Lorem ipsum dolor sit amet","description_job": "Lorem ipsum dolor sit amet","description_hobbies": "Lorem ipsum dolor sit amet","description_motto": "Lorem ipsum dolor sit amet","picture_thumb_url": "https://youthweb.net/img/steckbriefe/default_pic_m.jpg","picture_url": "https://youthweb.net/img/steckbriefe/default_pic_m.jpg"}}}');
        $userResponse->method('getHeader')->willReturn(['content-type' => 'json']);
        $userResponse->method('getStatusCode')->willReturn(200);

        $client = $this->createMock('GuzzleHttp\ClientInterface');
        $client->expects($this->exactly(2))
            ->method('send')
            ->willReturn($postResponse, $userResponse);
        $this->provider->setHttpClient($client);

        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
        $user = $this->provider->getResourceOwner($token);

        $this->assertEquals($userId, $user->getId());
        $this->assertEquals($userId, $user->toArray()['data']['id']);
        $this->assertEquals($firstname . ' ' . $lastname, $user->getName());
        $this->assertEquals($firstname . ' ' . $lastname, $user->toArray()['data']['attributes']['first_name'] . ' ' . $user->toArray()['data']['attributes']['last_name']);
        $this->assertEquals($nickname, $user->getNickname());
        $this->assertEquals($nickname, $user->toArray()['data']['attributes']['username']);
        $this->assertEquals($email, $user->getEmail());
        $this->assertEquals($email, $user->toArray()['data']['attributes']['email']);
        $this->assertStringContainsString((string) $userId, $user->getUrl());
    }

    /**
     * @test errors
     **/
    public function testExceptionThrownWhenErrorObjectWithDetailReceived()
    {
        $status = rand(400, 600);
        $postResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
        $postResponse->method('getBody')->willReturn('{"errors": [{"status": "406","title": "Not Acceptable","detail": "You havn\'t specified the the Accept Header. You have to use Accept application/vnd.api+json, application/vnd.api+json; net.youthweb.api.version="}]}');
        $postResponse->method('getHeader')->willReturn(['content-type' => 'json']);
        $postResponse->method('getStatusCode')->willReturn($status);

        $client = $this->createMock('GuzzleHttp\ClientInterface');
        $client->expects($this->once())
            ->method('send')
            ->willReturn($postResponse);
        $this->provider->setHttpClient($client);

        $this->expectException('League\OAuth2\Client\Provider\Exception\IdentityProviderException');
        $this->expectExceptionMessage('You havn\'t specified the the Accept Header. You have to use Accept application/vnd.api+json, application/vnd.api+json; net.youthweb.api.version=');

        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
    }

    /**
     * @test errors
     **/
    public function testExceptionThrownWhenErrorObjectWithoutDetailReceived()
    {
        $status = rand(400, 600);
        $postResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
        $postResponse->method('getBody')->willReturn('{"errors": [{"status": "406","title": "Not Acceptable"}]}');
        $postResponse->method('getHeader')->willReturn(['content-type' => 'json']);
        $postResponse->method('getStatusCode')->willReturn($status);

        $client = $this->createMock('GuzzleHttp\ClientInterface');
        $client->expects($this->once())
            ->method('send')
            ->willReturn($postResponse);
        $this->provider->setHttpClient($client);

        $this->expectException('League\OAuth2\Client\Provider\Exception\IdentityProviderException');
        $this->expectExceptionMessage('Not Acceptable');

        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
    }

    /**
     * @test errors
     **/
    public function testExceptionThrownWhenOAuthErrorReceived()
    {
        $status = rand(400, 600);
        $postResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
        $postResponse->method('getBody')->willReturn('{"error": "invalid_request","error_description": "The request is missing a required parameter, includes an unsupported parameter value (other than grant type), repeats a parameter, includes multiple credentials, utilizes more than one mechanism for authenticating the client, or is otherwise malformed.","error_uri": "https://tools.ietf.org/html/rfc6749#section-5.2"}');
        $postResponse->method('getHeader')->willReturn(['content-type' => 'json']);
        $postResponse->method('getStatusCode')->willReturn($status);

        $client = $this->createMock('GuzzleHttp\ClientInterface');
        $client->expects($this->once())
            ->method('send')
            ->willReturn($postResponse);
        $this->provider->setHttpClient($client);

        $this->expectException('League\OAuth2\Client\Provider\Exception\IdentityProviderException');
        $this->expectExceptionMessage('invalid_request');

        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
    }
}


tests/unit/Provider/YouthwebTest.php
<?php

namespace Youthweb\OAuth2\Client\Test\Provider;

class YouthwebTest extends \PHPUnit_Framework_TestCase
{
	protected $provider;

	protected function setUp()
	{
		$this->provider = new \Youthweb\OAuth2\Client\Provider\Youthweb([
			'clientId' => 'mock_client_id',
			'clientSecret' => 'mock_secret',
			'redirectUri' => 'none',
		]);
	}

	public function testAuthorizationUrl()
	{
		$url = $this->provider->getAuthorizationUrl();
		$uri = parse_url($url);
		parse_str($uri['query'], $query);

		$this->assertArrayHasKey('client_id', $query);
		$this->assertArrayHasKey('redirect_uri', $query);
		$this->assertArrayHasKey('state', $query);
		$this->assertArrayHasKey('scope', $query);
		$this->assertArrayHasKey('response_type', $query);
		$this->assertArrayHasKey('approval_prompt', $query);
		$this->assertNotNull($this->provider->getState());
	}

	public function testScopes()
	{
		$options = ['scope' => [uniqid(),uniqid()]];

		$url = $this->provider->getAuthorizationUrl($options);

		$this->assertContains(urlencode(implode(',', $options['scope'])), $url);
	}

	public function testGetAuthorizationUrl()
	{
		$url = $this->provider->getAuthorizationUrl();
		$uri = parse_url($url);

		$this->assertEquals('/auth/authorize', $uri['path']);
	}

	public function testGetBaseAccessTokenUrl()
	{
		$params = [];

		$url = $this->provider->getBaseAccessTokenUrl($params);
		$uri = parse_url($url);

		$this->assertEquals('/auth/access_token', $uri['path']);
	}

	public function testGetAccessToken()
	{
		$response = $this->createMock('Psr\Http\Message\ResponseInterface');
		$response->method('getBody')->willReturn('{"access_token":"mock_access_token", "scope":"user:read", "token_type":"bearer"}');
		$response->method('getHeader')->willReturn(['content-type' => 'json']);
		$response->method('getStatusCode')->willReturn(200);

		$client = $this->createMock('GuzzleHttp\ClientInterface');
		$client->expects($this->once())->method('send')->willReturn($response);
		$this->provider->setHttpClient($client);

		$token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);

		$this->assertEquals('mock_access_token', $token->getToken());
		$this->assertNull($token->getExpires());
		$this->assertNull($token->getRefreshToken());
		$this->assertNull($token->getResourceOwnerId());
	}

	public function testUserData()
	{
		$userId = rand(1000,99999);
		$firstname = uniqid();
		$lastname = uniqid();
		$nickname = uniqid();
		$email = uniqid();

		$postResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
		$postResponse->method('getBody')->willReturn('access_token=mock_access_token&expires=3600&refresh_token=mock_refresh_token&otherKey={1234}');
		$postResponse->method('getHeader')->willReturn(['content-type' => 'application/x-www-form-urlencoded']);
		$postResponse->method('getStatusCode')->willReturn(200);

		$userResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
		$userResponse->method('getBody')->willReturn('{"data": {"type": "users","id": "'.$userId.'","attributes": {"username": "'.$nickname.'","first_name": "'.$firstname.'","last_name": "'.$lastname.'","email": "'.$email.'","birthday": "1988-03-05","created_at": "2006-01-01T21:00:00+01:00","last_login": "2016-01-01T22:00:00+02:00","zip": "12345","city": "Jamestown","description_jesus": "Lorem ipsum dolor sit amet","description_job": "Lorem ipsum dolor sit amet","description_hobbies": "Lorem ipsum dolor sit amet","description_motto": "Lorem ipsum dolor sit amet","picture_thumb_url": "https://youthweb.net/img/steckbriefe/default_pic_m.jpg","picture_url": "https://youthweb.net/img/steckbriefe/default_pic_m.jpg"}}}');
		$userResponse->method('getHeader')->willReturn(['content-type' => 'json']);
		$userResponse->method('getStatusCode')->willReturn(200);

		$client = $this->createMock('GuzzleHttp\ClientInterface');
		$client->expects($this->exactly(2))
			->method('send')
			->willReturn($postResponse, $userResponse);
		$this->provider->setHttpClient($client);

		$token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
		$user = $this->provider->getResourceOwner($token);

		$this->assertEquals($userId, $user->getId());
		$this->assertEquals($userId, $user->toArray()['data']['id']);
		$this->assertEquals($firstname.' '.$lastname, $user->getName());
		$this->assertEquals($firstname.' '.$lastname, $user->toArray()['data']['attributes']['first_name'].' '.$user->toArray()['data']['attributes']['last_name']);
		$this->assertEquals($nickname, $user->getNickname());
		$this->assertEquals($nickname, $user->toArray()['data']['attributes']['username']);
		$this->assertEquals($email, $user->getEmail());
		$this->assertEquals($email, $user->toArray()['data']['attributes']['email']);
		$this->assertContains((string) $userId, $user->getUrl());
	}

	/**
	 * @test errors
	 **/
	public function testExceptionThrownWhenErrorObjectWithDetailReceived()
	{
		$status = rand(400,600);
		$postResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
		$postResponse->method('getBody')->willReturn('{"errors": [{"status": "406","title": "Not Acceptable","detail": "You havn\'t specified the the Accept Header. You have to use Accept application/vnd.api+json, application/vnd.api+json; net.youthweb.api.version="}]}');
		$postResponse->method('getHeader')->willReturn(['content-type' => 'json']);
		$postResponse->method('getStatusCode')->willReturn($status);

		$client = $this->createMock('GuzzleHttp\ClientInterface');
		$client->expects($this->once())
			->method('send')
			->willReturn($postResponse);
		$this->provider->setHttpClient($client);

		$this->setExpectedException(
			'League\OAuth2\Client\Provider\Exception\IdentityProviderException',
			'You havn\'t specified the the Accept Header. You have to use Accept application/vnd.api+json, application/vnd.api+json; net.youthweb.api.version='
		);

		$token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
	}

	/**
	 * @test errors
	 **/
	public function testExceptionThrownWhenErrorObjectWithoutDetailReceived()
	{
		$status = rand(400,600);
		$postResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
		$postResponse->method('getBody')->willReturn('{"errors": [{"status": "406","title": "Not Acceptable"}]}');
		$postResponse->method('getHeader')->willReturn(['content-type' => 'json']);
		$postResponse->method('getStatusCode')->willReturn($status);

		$client = $this->createMock('GuzzleHttp\ClientInterface');
		$client->expects($this->once())
			->method('send')
			->willReturn($postResponse);
		$this->provider->setHttpClient($client);

		$this->setExpectedException(
			'League\OAuth2\Client\Provider\Exception\IdentityProviderException',
			'Not Acceptable'
		);

		$token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
	}

	/**
	 * @test errors
	 **/
	public function testExceptionThrownWhenOAuthErrorReceived()
	{
		$status = rand(400,600);
		$postResponse = $this->createMock('Psr\Http\Message\ResponseInterface');
		$postResponse->method('getBody')->willReturn('{"error": "invalid_request","error_description": "The request is missing a required parameter, includes an unsupported parameter value (other than grant type), repeats a parameter, includes multiple credentials, utilizes more than one mechanism for authenticating the client, or is otherwise malformed.","error_uri": "https://tools.ietf.org/html/rfc6749#section-5.2"}');
		$postResponse->method('getHeader')->willReturn(['content-type' => 'json']);
		$postResponse->method('getStatusCode')->willReturn($status);

		$client = $this->createMock('GuzzleHttp\ClientInterface');
		$client->expects($this->once())
			->method('send')
			->willReturn($postResponse);
		$this->provider->setHttpClient($client);

		$this->setExpectedException(
			'League\OAuth2\Client\Provider\Exception\IdentityProviderException',
			'invalid_request'
		);

		$token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
	}
}


src/Provider/YouthwebResourceOwner.php
<?php
/*
 * A PHP Library to simplify the OAuth2 process with youthweb.net.
 * Copyright (C) 2016-2019  Youthweb e.V. <info@youthweb.net>

 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Youthweb\OAuth2\Client\Provider;

use League\OAuth2\Client\Provider\ResourceOwnerInterface;

class YouthwebResourceOwner implements ResourceOwnerInterface
{
    /**
     * Domain
     *
     * @var string
     */
    protected $domain;

    /**
     * Raw response
     *
     * @var array
     */
    protected $response;

    /**
     * Creates new resource owner.
     *
     * @param array $response
     */
    public function __construct(array $response = [])
    {
        $this->response = $response;
    }

    /**
     * Get resource owner id
     *
     * @return string|null
     */
    public function getId()
    {
        return $this->response['data']['id'] ?: null;
    }

    /**
     * Get resource owner email
     *
     * @return string|null
     */
    public function getEmail()
    {
        return $this->response['data']['attributes']['email'] ?: null;
    }

    /**
     * Get resource owner name
     *
     * @return string|null
     */
    public function getName()
    {
        $name = null;

        if (isset($this->response['data']['attributes']['first_name'])) {
            $name = $this->response['data']['attributes']['first_name'] . ' ' . $this->response['data']['attributes']['last_name'];
        }

        return $name;
    }

    /**
     * Get resource owner nickname
     *
     * @return string|null
     */
    public function getNickname()
    {
        return $this->response['data']['attributes']['username'] ?: null;
    }

    /**
     * Get resource owner url
     *
     * @return string|null
     */
    public function getUrl()
    {
        return trim($this->domain . '/users/' . $this->getId()) ?: null;
    }

    /**
     * Set resource owner domain
     *
     * @param string $domain
     *
     * @return ResourceOwner
     */
    public function setDomain($domain)
    {
        $this->domain = $domain;

        return $this;
    }

    /**
     * Return all of the owner details available as an array.
     *
     * @return array
     */
    public function toArray()
    {
        return $this->response;
    }
}


src/Provider/Exception/YouthwebIdentityProviderException.php
<?php
/*
 * A PHP Library to simplify the OAuth2 process with youthweb.net.
 * Copyright (C) 2016-2019  Youthweb e.V. <info@youthweb.net>

 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Youthweb\OAuth2\Client\Provider\Exception;

use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
use Psr\Http\Message\ResponseInterface;

class YouthwebIdentityProviderException extends IdentityProviderException
{
    /**
     * Creates client exception from response.
     *
     * @param ResponseInterface $response
     * @param string            $data     Parsed response data
     *
     * @return YouthwebIdentityProviderException
     */
    public static function clientException(ResponseInterface $response, $data)
    {
        $message = $response->getReasonPhrase();

        // Use only the first error
        if (isset($data['errors'][0]['detail'])) {
            $message = (string) $data['errors'][0]['detail'];
        } elseif (isset($data['errors'][0]['title'])) {
            $message = (string) $data['errors'][0]['title'];
        }

        return static::fromResponse($response, $message);
    }

    /**
     * Creates oauth exception from response.
     *
     * @param ResponseInterface $response
     * @param string            $data     Parsed response data
     *
     * @return YouthwebIdentityProviderException
     */
    public static function oauthException(ResponseInterface $response, $data)
    {
        return static::fromResponse(
            $response,
            isset($data['error']) ? $data['error'] : $response->getReasonPhrase()
        );
    }

    /**
     * Creates identity exception from response.
     *
     * @param ResponseInterface $response
     * @param string            $message
     *
     * @return YouthwebIdentityProviderException
     */
    protected static function fromResponse(ResponseInterface $response, $message = null)
    {
        return new static($message, $response->getStatusCode(), (string) $response->getBody());
    }
}


src/Provider/Youthweb.php
<?php
/*
 * A PHP Library to simplify the OAuth2 process with youthweb.net.
 * Copyright (C) 2016-2019  Youthweb e.V. <info@youthweb.net>

 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Youthweb\OAuth2\Client\Provider;

use Youthweb\OAuth2\Client\Provider\Exception\YouthwebIdentityProviderException;
use League\OAuth2\Client\Provider\AbstractProvider;
use League\OAuth2\Client\Token\AccessToken;
use League\OAuth2\Client\Tool\BearerAuthorizationTrait;
use Psr\Http\Message\ResponseInterface;

class Youthweb extends AbstractProvider
{
    use BearerAuthorizationTrait;

    /**
     * Domain
     *
     * @var string
     */
    public $domain = 'https://youthweb.net';

    /**
     * Api domain
     *
     * @var string
     */
    public $apiDomain = 'https://api.youthweb.net';
    /**
     * Api version
     *
     * @var string
     */
    public $apiVersion = '0.18';

    /**
     * Get authorization url to begin OAuth flow
     *
     * @return string
     */
    public function getBaseAuthorizationUrl()
    {
        return $this->domain . '/auth/authorize';
    }

    /**
     * Get access token url to retrieve token
     *
     * @param array $params
     *
     * @return string
     */
    public function getBaseAccessTokenUrl(array $params)
    {
        return $this->domain . '/auth/access_token';
    }

    /**
     * Get provider url to fetch user details
     *
     * @param AccessToken $token
     *
     * @return string
     */
    public function getResourceOwnerDetailsUrl(AccessToken $token)
    {
        return $this->apiDomain . '/me';
    }

    /**
     * Returns an authenticated PSR-7 request instance for Youthweb-API.
     *
     * @param string             $method
     * @param string             $url
     * @param AccessToken|string $token
     * @param array              $options any of "headers", "body", and "protocolVersion"
     *
     * @return RequestInterface
     */
    public function getAuthenticatedRequest($method, $url, $token, array $options = [])
    {
        $options['headers']['Accept'] = 'application/vnd.api+json, application/vnd.api+json; net.youthweb.api.version=' . $this->apiVersion;
        $options['headers']['Content-Type'] = 'application/vnd.api+json';

        return parent::getAuthenticatedRequest($method, $url, $token, $options);
    }

    /**
     * Get the default scopes used by this provider.
     *
     * This should not be a complete list of all scopes, but the minimum
     * required for the provider user interface!
     *
     * @return array
     */
    protected function getDefaultScopes()
    {
        return [];
    }

    /**
     * Check a provider response for errors.
     *
     * @see   http://jsonapi.org/format/1.0/#errors
     * @see   https://tools.ietf.org/html/rfc6749#section-5.2
     *
     * @param ResponseInterface $response
     * @param string            $data     Parsed response data
     *
     * @throws IdentityProviderException
     */
    protected function checkResponse(ResponseInterface $response, $data)
    {
        if ($response->getStatusCode() >= 400) {
            // check for JSON API errors
            if (isset($data['errors'])) {
                throw YouthwebIdentityProviderException::clientException($response, $data);
            }

            // It must be an oauth2 error
            throw YouthwebIdentityProviderException::oauthException($response, $data);
        }
    }

    /**
     * Generate a user object from a successful user details request.
     *
     * @param array       $response
     * @param AccessToken $token
     *
     * @return League\OAuth2\Client\Provider\ResourceOwnerInterface
     */
    protected function createResourceOwner(array $response, AccessToken $token)
    {
        $user = new YouthwebResourceOwner($response);

        return $user->setDomain($this->domain);
    }
}
